testing/integration/beats_serverless_test.go

```diff
@@ -0,0 +1,589 @@
+)
+
+type BeatRunner struct {
+	suite.Suite
+	requirementsInfo *define.Info
+	agentFixture     *atesting.Fixture
+
+	// connection info
+	ESHost  string
+	user    string
+	pass    string
+	kibHost string
+
+	testUuid     string
+	testbeatName string
+}
+
+func TestBeatsServerless(t *testing.T) {
+	info := define.Require(t, define.Requirements{
+		OS: []define.OS{
+			{Type: define.Linux},
+		},
+		Stack: &define.Stack{},
+		Local: false,
+		Sudo:  true,
+	})
+
+	suite.Run(t, &BeatRunner{requirementsInfo: info})
+}
+
+func (runner *BeatRunner) SetupSuite() {
+	runner.T().Logf("In SetupSuite")
+
+	runner.testbeatName = os.Getenv("TEST_BINARY_NAME")
+	if runner.testbeatName == "" {
+		runner.T().Fatalf("TEST_BINARY_NAME must be set")
+	}
+	if runner.testbeatName == "elastic-agent" {
+		runner.T().Skipf("tests must be run against a beat, not elastic-agent")
+	}
+
+	if runner.testbeatName != "filebeat" && runner.testbeatName != "metricbeat" {
+		runner.T().Skip("test only supports metricbeat or filebeat")
+	}
+	runner.T().Logf("running serverless tests with %s", runner.testbeatName)
+
+	agentFixture, err := define.NewFixtureWithBinary(runner.T(), define.Version(), runner.testbeatName, "/home/ubuntu", atesting.WithRunLength(time.Minute*3), atesting.WithAdditionalArgs([]string{"-E", "output.elasticsearch.allow_older_versions=true"}))
+	runner.agentFixture = agentFixture
+	require.NoError(runner.T(), err)
+
+	// the require.* code will fail without these, so assume the values are non-nil
+	runner.ESHost = os.Getenv("ELASTICSEARCH_HOST")
+	runner.user = os.Getenv("ELASTICSEARCH_USERNAME")
+	runner.pass = os.Getenv("ELASTICSEARCH_PASSWORD")
+	runner.kibHost = os.Getenv("KIBANA_HOST")
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+
+	beatOutConfig := `
+output.elasticsearch:
+  hosts: ["%s"]
+  api_key: "%s:%s"
+setup.kibana:
+  host: %s
+metricbeat.config.modules:
+  path: ${path.config}/modules.d/*.yml
+processors:
+  - add_fields:
+      target: host
+      fields:
+        test-id: %s
+`
+	if runner.testbeatName == "filebeat" {
+		beatOutConfig = `
+output.elasticsearch:
+  hosts: ["%s"]
+  api_key: "%s:%s"
+setup.kibana:
+  host: %s
+filebeat.config.modules:
+  - modules: system
+    syslog:
+      enabled: true
+    auth:
+      enabled: true
+processors:
+  - add_fields:
+      target: host
+      fields:
+        test-id: %s
+`
+	}
+
+	apiResp, err := estools.CreateAPIKey(ctx, runner.requirementsInfo.ESClient, estools.APIKeyRequest{Name: "test-api-key", Expiration: "1d"})
+	require.NoError(runner.T(), err)
+
+	// beats likes to add standard ports to URLs that don't have them, and ESS will sometimes return a URL without a port, assuming :443
+	// so try to fix that here
+	fixedKibanaHost := runner.kibHost
+	parsedKibana, err := url.Parse(runner.kibHost)
+	require.NoError(runner.T(), err)
+	if parsedKibana.Port() == "" {
+		fixedKibanaHost = fmt.Sprintf("%s:443", fixedKibanaHost)
+	}
```

Comment:
Shouldn't this conditioned to the url scheme being https (which is 99.9% of the time is)? It's just that setting fixed `:443` to a URL where we don't even check the scheme feels wrong...
Furthermore, since we spent the time parsing the URL couldn't we set `parsedKibana.Host` field and then call `parsedKibanaHost.String()` ?

Reply:
You would think, but the beats defaults to `5601` if you don't specify the port, hence why we need to add it manually.